### PR TITLE
New rule S3981 Collection sizes and array length comparisons should make sense

### LINKED ISSUE #596: Rule: Collection sizes and array length comparisons should make sense
Implements [RSPEC-3981](https://jira.sonarsource.com/browse/RSPEC-3981)
common rule

### PR DESCRIPTION
**Pull request by saberduck**
Fixes #596

### CODE REVIEW

File: javascript-checks/src/main/java/org/sonar/javascript/checks/CollectionSizeAndArrayLengthCheck.java [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+      }
+    }
+    super.visitBinaryExpression(tree);
+  }
+
+  private String getPropertyName(MemberExpressionTree memberExpressionTree) {
+    return ((IdentifierTree) memberExpressionTree.property()).name();
+  }
+
+  private boolean isLengthOrSizeProperty(DotMemberExpressionTree memberExpressionTree) {
```
Comment (SonarTech) [inline javascript-checks/src/main/java/org/sonar/javascript/checks/CollectionSizeAndArrayLengthCheck.java]: Code Smell Code Smell: Make "isLengthOrSizeProperty" a "static" method. ([squid:S2325](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2325))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.sonarsource.javascript%3Ajavascript&branch=issue-596&issues=AWFNPTBTNbtQu1yJ826f&open=AWFNPTBTNbtQu1yJ826f)

[ISSUE_KEY]: AWFNPTBTNbtQu1yJ826f

File: javascript-checks/src/main/java/org/sonar/javascript/checks/CollectionSizeAndArrayLengthCheck.java [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+    String name = memberExpressionTree.property().name();
+    return "length".equals(name) || "size".equals(name);
+  }
+
+  private boolean isZeroLiteral(ExpressionTree tree) {
```
Comment (SonarTech) [inline javascript-checks/src/main/java/org/sonar/javascript/checks/CollectionSizeAndArrayLengthCheck.java]: Code Smell Code Smell: Make "isZeroLiteral" a "static" method. ([squid:S2325](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2325))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.sonarsource.javascript%3Ajavascript&branch=issue-596&issues=AWFNPTBTNbtQu1yJ826g&open=AWFNPTBTNbtQu1yJ826g)

[ISSUE_KEY]: AWFNPTBTNbtQu1yJ826g

File: javascript-checks/src/main/java/org/sonar/javascript/checks/CollectionSizeAndArrayLengthCheck.java [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+      ExpressionTree leftOperand = tree.leftOperand();
+      if (leftOperand.is(Tree.Kind.DOT_MEMBER_EXPRESSION)
+          && isLengthOrSizeProperty((DotMemberExpressionTree) leftOperand)) {
+        addIssue(tree, String.format("The %s is always \">=0\", so fix this test to get the real expected behavior.", getPropertyName((MemberExpressionTree) leftOperand)));
+      }
+    }
+    super.visitBinaryExpression(tree);
+  }
+
+  private String getPropertyName(MemberExpressionTree memberExpressionTree) {
```
Comment (SonarTech) [inline javascript-checks/src/main/java/org/sonar/javascript/checks/CollectionSizeAndArrayLengthCheck.java]: Code Smell Code Smell: Make "getPropertyName" a "static" method. ([squid:S2325](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2325))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.sonarsource.javascript%3Ajavascript&branch=issue-596&issues=AWFNPTBTNbtQu1yJ826e&open=AWFNPTBTNbtQu1yJ826e)

[ISSUE_KEY]: AWFNPTBTNbtQu1yJ826e

File: javascript-checks/src/main/java/org/sonar/javascript/checks/CollectionSizeAndArrayLengthCheck.java [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+import org.sonar.check.Rule;
+import org.sonar.plugins.javascript.api.tree.Tree;
+import org.sonar.plugins.javascript.api.tree.expression.BinaryExpressionTree;
+import org.sonar.plugins.javascript.api.tree.expression.DotMemberExpressionTree;
+import org.sonar.plugins.javascript.api.tree.expression.ExpressionTree;
+import org.sonar.plugins.javascript.api.tree.expression.LiteralTree;
+import org.sonar.plugins.javascript.api.visitors.DoubleDispatchVisitorCheck;
+
+@Rule(key = "S3981")
+public class CollectionSizeAndArrayLengthCheck extends DoubleDispatchVisitorCheck {
```
Comment (vilchik-elena) [inline javascript-checks/src/main/java/org/sonar/javascript/checks/CollectionSizeAndArrayLengthCheck.java]: @saberduck minor: can we rename this check to something shorter?) e.g.`CollectionSizeComparisonCheck`?

File: javascript-checks/src/main/java/org/sonar/javascript/checks/CollectionSizeAndArrayLengthCheck.java [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+
+  @Override
+  public void visitBinaryExpression(BinaryExpressionTree tree) {
+    if (tree.is(Tree.Kind.LESS_THAN, Tree.Kind.GREATER_THAN_OR_EQUAL_TO) && isZeroLiteral(tree.rightOperand())) {
+      ExpressionTree leftOperand = tree.leftOperand();
+      if (leftOperand.is(Tree.Kind.DOT_MEMBER_EXPRESSION)
+          && isLengthOrSizeProperty((DotMemberExpressionTree) leftOperand)) {
```
Comment (vilchik-elena) [inline javascript-checks/src/main/java/org/sonar/javascript/checks/CollectionSizeAndArrayLengthCheck.java]: @saberduck is this new line required?

File: javascript-checks/src/test/resources/checks/collectionSizeAndArrayLength.js [hunk 6]
```diff
@@ -0,0 +1,25 @@
+let arr = [];
+let mySet = new Set();
+let myMap = new Map();
+
+
+if (mySet.size < 0) { } // Noncompliant {{The size is always ">=0", so fix this test to get the real expected behavior.}}
```
Comment (vilchik-elena) [inline javascript-checks/src/test/resources/checks/collectionSizeAndArrayLength.js]: @saberduck Let's make the message the same as for TS:
`The %s of "%s" is always ">=0", so fix this test to get the real expected behavior.`

File: javascript-checks/src/test/resources/checks/collectionSizeAndArrayLength.js [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+
+// OK
+
+if (arr.length < 1) { }
+if (arr.length > 0) { }
+if (arr.length <= 1) { }
+if (arr.length >= 1) { }
+if (arr.length < 50) { }
+if (arr.length < 5 + 0) { }
+if (obj.size() >= 0) { }
```
Comment (vilchik-elena) [inline javascript-checks/src/test/resources/checks/collectionSizeAndArrayLength.js]: @saberduck add test with complex expression
`foo.bar().length >= 0`

File: javascript-checks/src/test/resources/checks/collectionSizeAndArrayLength.js [hunk 8]
```diff
@@ -0,0 +1,25 @@
+let arr = [];
```
Comment (vilchik-elena) [inline javascript-checks/src/test/resources/checks/collectionSizeAndArrayLength.js]: @saberduck I prefer to drop these declarations to be explicit that this rules doesn't rely on any semantic information. Variable names are enough to get the point of the rule

File: javascript-checks/src/test/resources/checks/collectionSizeAndArrayLength.js (hunk 7 above)
Comment (vilchik-elena) [reply javascript-checks/src/test/resources/checks/collectionSizeAndArrayLength.js]: @saberduck add test with FP:
```
let obj = { };
obj.length = -42;
if (obj.length < 0) {} // Noncompliant, FP
```

**Comment by saberduck** [comment]
@vilchik-elena I applied your comments, please have another look

File: javascript-checks/src/test/resources/checks/collectionSize.js [hunk 9]
```diff
@@ -0,0 +1,27 @@
+
```
Comment (vilchik-elena) [inline javascript-checks/src/test/resources/checks/collectionSize.js]: @saberduck I prefer to have the name of test file same as check (easier to find)

File: javascript-checks/src/test/resources/checks/collectionSize.js [hunk 10]
```diff
@@ -0,0 +1,27 @@
...
+if (mySet.size < 0) { } // Noncompliant {{Fix this expression; size of "mySet" is always greater or equal to zero.}}
+//  ^^^^^^^^^^^^^^
+if (myMap.size < 0) { } // Noncompliant {{Fix this expression; size of "myMap" is always greater or equal to zero.}}
```
Comment (vilchik-elena) [inline javascript-checks/src/test/resources/checks/collectionSize.js]: @saberduck you know, you don't have to assert every message:)

**Comment by vilchik-elena** [comment]
@saberduck Feel free to merge (after renaming test file:)) 
I've tested manually

**Pull request merged**